File: src/Coachable/SelectionSort.java
package Coachable;

import java.util.Arrays;


public class SelectionSort {
	public static void main(String[] args) {

		//int[] arr = new int[]{-20, -10, 30, 0, -1, 4};
		int[] arr = new int[]{4,0,18,24,16,20,4,18,19,8,14,13};

		selectionSorting(arr);

		System.out.println(Arrays.toString(arr));


		for (int num : arr){
			char c = (char) (num + 65);
			System.out.print(c + " ");
		}
	}

	/**
	 * Selection Sorting
	 * Time Complexity -> O(n^2)
	 * Space Complexity -> O(1)
	 */

	private static void selectionSorting(int[] arr) {
		for (int i = 0; i < arr.length; i++) {
			int curr = arr[i];
			int minIdx = i;
			for (int j = i; j < arr.length; j++){
				if (arr[j] < arr[minIdx])
					minIdx = j;
			}
			int temp = arr[i];
			arr[i] = arr[minIdx];
			arr[minIdx] = temp;
		}
	}

}

File: src/Coachable/1-DSA/4-String,Sorting/SelectionSort.py
"""
Selection Sort
"""

"""
Elementary Sort

Stable sorting
Time: O(n^2)
Space: O(1)
"""


class SelectionSort:

    @staticmethod
    def __init__(self):
        pass

    @staticmethod
    def sort(array):
        i, n = 1, len(array)

        for i in range(n - 1):
            curr_smallest, curr_smallest_idx = array[i], i
            for j in range(i + 1, n):
                if array[j] < curr_smallest:
                    curr_smallest, curr_smallest_idx = array[j], j
            array[i], array[curr_smallest_idx] = array[curr_smallest_idx], array[i]

        return array


arr = [8, -5, 10, 2, 1]
SelectionSort.sort(arr)
print(arr)
